### ISSUE TITLE
ProductListingPriceIndexer causes n+1 db requests

### ISSUE THREAD

**Issue by htuscher** [shopware/shopware#691]
**Description**  
We're trying to do heavy price updates using the API. For example we want to update the prices of 1000 products at once.  To simplify things, we built our own API endpoint, but that doesn't matter much.

Basically this is executed:
```php
$result = $this->productPriceRepository->upsert($this->upserts, Context::createDefaultContext());
```

`$this->upserts` contain 1000 operations in this form:
```php
        $this->upserts[] = [
            'id'            => $priceId,
            'productId'     => $productId,
            'ruleId'        => $ruleId,
            'quantityStart' => 1,
            'price'         => [
                [
                    'currencyId' => $currencyId,
                    'net'        => $net,
                    'gross'      => $gross,
                    'linked'     => true,
                ],
            ],
        ];
```

The API request is damn slow (35s) on a heavy machine.

Therefore I profiled the execution using Tideways. This was the outcome:
For 1000 articles there are over 9000 queries executed. Following the callgraph, only 1000 were caused by the actual repository write. 

2000 queries were executed by the `ProductListingPriceIndexer::update` which took 70% of the total execution time.
An additional 2000 queries were executed by `VariantListingIndexer::update` which took 10% of the total execution time.

But let's focus on the `ProductListingPriceIndexer::update`.  It fetches all prices for a product and updates the listing_prices. If you have e.g. 1000 prices for an article (rules by customer), then you would trigger for a single price update 1000 listing price updates.

**Possible Solution**  

This should definitely be bound to the prices, not the products. As a price belongs only to one rule. The other prices for the other rules don't need updating at this point.

Therefore the `ProductListingIndexer` should only fetch the prices for the rules for which the prices have been updated.

**Comment by mitelg** [comment]
hey @hhoechtl 
we have refactored the indexer in the current master branch, which will be the 6.2 release in the near future. could you please check, if you have your problems there, too?

**Comment by htuscher** [comment]
@mitelg I just checked on `master` and it got a lot better. But you're still recalculating the prices for all rules, even though I'm just modifying the price for 1 rule. Is this really necessary? Wouldn't it be faster to load the `listing_prices` field from the `product` table, decode the content and replace only the structs and `formatted` prices by key?

Also would it be possible to move the entire indexing process to a message that could be handled asynchronously?
Let's say, we update 1000 product prices in batches by rule. It would be sufficient to run reindexing only once after all batches are processed (scheduler or messages). That would safe up to 90% of computing time.

**Comment by OliverSkroblin** [comment]
hey @hhoechtl,

you can use the sync api to update the prices. If you provide the header `message-queue-indexing` the indexing will be done over message queue.

**Comment by htuscher** [comment]
Thx @OliverSkroblin for the tip with the header, I'll take a look. That will at least make the client terminate the connection faster. 

**Nevertheless the calculation overhead for untouched prices remains the same.** 

We have built our own API because of https://issues.shopware.com/issues/NEXT-5432. We update Prices by productNumber and customerNumber (which automatically generates a customer specific rule if not present).

**Comment by shyim** [comment]
The indexing has been rewritten in the meanwhile. Feel free to create a new issue if you have still problems
